### PR TITLE
'classes' body support for custom promises

### PR DESCRIPTION
**Pull request by vpodzime**
Steps:
- [x] implementation
- [x] acceptance tests

### CODE REVIEW

**Comment by vpodzime** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6234/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6234/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6234/

File: libpromises/mod_custom.c [hunk 1]
```diff
@@ -590,6 +591,11 @@ static inline bool CustomPromise_IsFullyResolved(const Promise *pp)
     for (size_t i = 0; i < attributes; i++)
     {
         const Constraint *attribute = SeqAt(pp->conlist, i);
+        if (IsClassesBodyConstraint(attribute->lval))
+        {
+            /* Not passed to the modules, handled on the agent side. */
+            continue;
+        }
```
Comment (olehermanse) [inline libpromises/mod_custom.c:598]: The way all of these are crammed into top level attributes is really unfortunate, but not your fault of course.

Please expand the list of attributes not sent to module in the spec here:

https://github.com/cfengine/core/blob/master/docs/custom_promise_types/modules.md#evaluation-details

File: libpromises/attributes.c [hunk 2]
```diff
@@ -31,6 +31,7 @@
 #include <chflags.h>
 #include <audit.h>
 #include <protocol.h> // ParseProtocolVersionPolicy()
+#include <string_lib.h>         /* StringEqual() */
```
Comment (olehermanse) [inline libpromises/attributes.c:34]: I'd prefer `//`, especially since it's used on line above.

**Review by olehermanse — APPROVED** [review]
Great improvement, thanks!

File: libpromises/mod_custom.c (hunk 1 above)
Comment (vpodzime) [reply libpromises/mod_custom.c:598]: > Please expand the list of attributes not sent to module in the spec here:

https://github.com/cfengine/core/pull/4552

**Pull request merged**